### PR TITLE
feat(new transform): Inital `kubernetes_pod_metadata` transform

### LINKED ISSUE #1072: New `kubernetes_pod_metadata` transform
We should provide a `kubernetes_pod_metadata` transform that is able to fetch a list of pods and enrich events. This should work based off of the kube openapi spec using the [`k8-openapi`] crate.

# Configuration

```toml
[transforms.my_transform_id]
  type = "kubernetes_pod_metadata"
  fields = ["instance-id", "local-hostname", "local-ipv4", "public-hostname", "public-ipv4"] # default
  namespace = ["your-namespace"] # optional, default is to list ALL pods in all namespaces
```

## Options

- `fields` contains the list of fields to include in each event.
- ~~`namespace` contains a list of namespaces that we should watch for a list of pods and their metadata.~~ See global schema for this field.

## Fields

We should provide this set of fields by default and they are overridable via the `fields` option.

- `pod_name`
- `pod_namespace`
- `pod_uid`
- `labels` (not sure if we should provide a way to include the label name in the key or have them as a tuple?)
- `annotations`
- `node_name`

Most of these data points should come from a combination of [`PodSpec`] and [`ObjectMeta`] using the [`Pod::list_namespaced_pod`].

[`k8-openapi`]: https://docs.rs/k8s-openapi
[`PodSpec`]: https://docs.rs/k8s-openapi/0.6.0/k8s_openapi/api/core/v1/struct.PodSpec.html
[`ObjectMeta`]: https://docs.rs/k8s-openapi/0.6.0/k8s_openapi/apimachinery/pkg/apis/meta/v1/struct.ObjectMeta.html
[`Pod::list_namespaced_pod`]: https://docs.rs/k8s-openapi/0.6.0/k8s_openapi/api/core/v1/struct.Pod.html#method.list_namespaced_pod

## Global Schema

`log_schema` should add a `kubernetes_namespace` field to correlate where nested fields should be added to events. This should replace the `namespace` transform option.

## Implementation

This should use the [`k8-openapi`] crate, plus the [`kube`] crate, though I would prefer we stick with just hyper and vendor the config loading. This implementation should follow the proposed method for implementing a stream table join transforms as seen in #1069.

[`kube`]: https://docs.rs/kube/0.16.1/kube/

## Questions

- How should we nest these? Should they live under a `kubernetes` key?
- How can we correlate pod events coming from the `kubernetes` source with the ones on the list? @ktff 

## Todo

- [x] `WatchClient`

- [ ] `kubernetes_pod_metadata`

### PR DESCRIPTION
**Pull request by ktff**
Closes #1072
Ref #1867

### Implemetation notes:
 * Supported fields:
   * `name`
   * `namespace`
   * `creation_timestamp`
   * `deletion_timestamp`
   * `labels`
   * `annotations`
   * `node_name`
   * `hostname`
   * `priority`
   * `priority_class_name`
   * `service_account_name`
   * `subdomain`
   * `host_ip`
   * `ip`
 * Fields are added under log schema `kubernetes_key`. Which has default value `kubernetes`.  
 * Tests and example config are using RBAC authorization to allow Vector access to API server. RBAC is usualy enabled by default, but this requirement should be documented .

### Open questions

- [x] Is default value `kubernetes` best for log schema `kubernetes_key` ?

- [x] Is this a breaking change because log schema has been changed ? (Edit: Will not be with #1923)

- [x] How to deal with `Event` interpreting label/annotation keys like "app.kubernetes.io/name" as nested? 
### Todo
 
- [x]  After this is merged, open an issue to update `kubernetes_source` naming scheme to use kubernetes_key. This will be a breaking change.

- [x] Add delete timeout.

- [x] Add jitter.

### CODE REVIEW

**Comment by binarylogic** [comment]
@ktff I added #1867 as a closing issue here. Please remove if that is not the case, but it looks like it at least partially resolves it?

**Comment by ktff** [comment]
> but it looks like it at least partially resolves it?

Yes, this two:
 
> - [ ] Add a log_schema.kubernetes_key option
> - [ ] Use this key in the kubernetes_metadata transform

**Comment by binarylogic** [comment]
@ktff would you mind adding docs?

Also, do we have a setting that controls the refresh rate against the k8s API? I ask because I have spoken to users that have experienced issues with this with fluent*. For example, very large clusters (5000+ pods) effectively DDOS'd the k8s API. A configuration option should allow users to reduce the refresh rate.

**Comment by ktff** [comment]
@binarylogic good question. We are currently only proactively asking for metadata once and then passively waiting for api server to send us changes, so we don't have refresh rate, but we have something similar.

We are using watch endpoint to listen for metadata changes for pods on local node, so there are basically two states:
 1. Connecting to api server.
    Exposing the `retry_timeout` as configuration option would enable users to decrease the pressure on api server in large clusters. Also, because 1. is problematic when large number of Vectors try to connect at the same time, deploying Vector with this transform on large clusters at once is problematic.
   To mitigate that, two more changes would help:
     - Make timeout duration probabilistic. Uniform distribution on range `[0,2*retry_timeout]` would have largest deviation, a median of `retry_timeout`, and also be indirectly depended on the size of the cluster. This would help spread out connection retries through time which would spread our pressure on api server more evenly. 
     - Add the timeout even when trying to connect for the first time. 

 2. Waiting for changes from api server.   
      Shouldn't be problematic because it's either self regulating or an issue that needs to be solved with extra feature when it arises.

It sounds reasonable to me to add this as it's a small change with great benefits.

@LucioFranco what do you think?

File: src/transforms/kubernetes/mod.rs [hunk 1]
```diff
@@ -1 +1,552 @@
...
+    MissingNodeName { env: &'static str },
+    #[snafu(display("Unknown metadata fields: {:?}", fields))]
+    UnknownFields { fields: Vec<String> },
+}
+
+/// Vlient which watches for Pod metadata changes, extracts fields,
+/// and writes them to the metadata map.
+struct MetadataClient {
+    fields: Vec<Box<dyn Fn(&Pod) -> Vec<(Atom, Value)> + Send + Sync + 'static>>,
+    metadata: WriteHandle<Bytes, Box<(Atom, FieldValue)>>,
```
Comment (LucioFranco) [inline src/transforms/kubernetes/mod.rs]: why does this need to be boxed?

File: src/transforms/kubernetes/mod.rs [hunk 2]
```diff
@@ -1 +1,552 @@
...
+
+pub struct KubernetesPodMetadata {
+    metadata: ReadHandle<Bytes, Box<(Atom, FieldValue)>>,
+}
+
+impl Transform for KubernetesPodMetadata {
+    fn transform(&mut self, mut event: Event) -> Option<Event> {
+        let log = event.as_mut_log();
+
+        if let Some(Value::Bytes(pod_uid)) = log.get(&POD_UID) {
```
Comment (LucioFranco) [inline src/transforms/kubernetes/mod.rs]: should we make the pod_uid an option that defaults to `POD_UID`?

File: src/transforms/kubernetes/mod.rs [hunk 3]
```diff
@@ -1 +1,552 @@
...
+        let rt = runtime();
+
+        assert!(config
+            .build(TransformContext::new_test(rt.executor()))
+            .is_err());
+    }
+}
+
+#[cfg(test)]
+mod tests {
```
Comment (LucioFranco) [inline src/transforms/kubernetes/mod.rs]: these should probably be called `integration_tests`

File: src/transforms/kubernetes/mod.rs [hunk 4]
```diff
@@ -1 +1,552 @@
...
+    };
+    (name, Box::new(fun) as Box<_>)
+}
+
+fn with_prefix(name: &str) -> String {
+    event::log_schema().kubernetes_key().as_ref().to_owned() + "." + name
+}
+
+#[cfg(test)]
+mod regular_tests {
```
Comment (LucioFranco) [inline src/transforms/kubernetes/mod.rs]: and this can just be called `tests`

**Review by LucioFranco — COMMENTED** [review]
Looking good I left a few comments inline.

File: src/transforms/kubernetes/mod.rs [hunk 5]
```diff
@@ -1 +1,552 @@
...
+/// 10min sounds a lot, but a metadata entry averages around 300B and in an extreme
+/// scenario when a pod is deleted every second, we would have ~200kB of probably
+/// not usefull data. Which is acceptable. The benefit of such a long delay
+/// is that we will certainly* have the metadata while we are processing the
+/// remaining logs from the deleted pod.
+const DELETE_DELAY: Duration = Duration::from_secs(10 * 60);
+
+#[derive(Deserialize, Serialize, Debug)]
+#[serde(deny_unknown_fields)]
+pub struct KubePodMetadata {
```
[the remark this answers is not in the document]
Comment (ktff) [reply src/transforms/kubernetes/mod.rs]: `retry_timeout` would, but `DELETE_DELAY` won't as it doesn't have an influence on the issue because it's only used after the api server is done with changing Pod metadata. 

And do you think that on large clusters deploying Vector with this transform could be an issue? 
[Related](https://github.com/timberio/vector/pull/1888#issuecomment-591079624)

File: src/transforms/kubernetes/mod.rs [hunk 6]
```diff
@@ -1 +1,552 @@
...
+        }
+    }
+
+    /// Listens for pod metadata changes and updates metadata map.
+    async fn run(mut self) -> BuildError {
+        let mut version = None;
+        let mut error = None;
+        loop {
+            // Build watcher stream
+            let mut watcher = match self.client.watch_metadata(version.clone(), error.take()) {
```
[the remark this answers is not in the document]
Comment (ktff) [reply src/transforms/kubernetes/mod.rs]: It would remove only that one match, but then it wouldn't be as clear that this function ends only when it errors. And on first glance Result would mislead that it can end in other ways.

File: src/transforms/kubernetes/mod.rs [hunk 7]
```diff
@@ -1 +1,552 @@
...
+
+#[typetag::serde(name = "kubernetes_pod_metadata")]
+impl TransformConfig for KubePodMetadata {
+    fn build(&self, cx: TransformContext) -> crate::Result<Box<dyn Transform>> {
+        // Main idea is to have a background task which will premptively
+        // acquire metadata for all pods on this node, and with it maintaine
+        // a map of extracted metadata.
+
+        // Construct WatchClient
+        let node = std::env::var(NODE_NAME_ENV)
```
[the remark this answers is not in the document]
Comment (ktff) [reply src/transforms/kubernetes/mod.rs]: No, we are setting it [here](https://github.com/timberio/vector/blob/c044571879671bb0416a8cca12a3d3b829682baa/config/kubernetes/vector-daemonset.yaml#L104). 

Config option sounds nice, but I would see it as a feature because, with it, this transform could be used outside Kubernetes cluster which would at least require adding an ability to load kube config file. I think users would expect/assume such ability.

File: src/transforms/kubernetes/mod.rs [hunk 8]
```diff
@@ -1 +1,552 @@
...
+            // Build watcher stream
+            let mut watcher = match self.client.watch_metadata(version.clone(), error.take()) {
+                Ok(watcher) => watcher,
+                Err(error) => return BuildError::WatchStreamBuild { source: error },
+            };
+            info!("Watching Pod metadata.");
+
+            // Watch loop
+            let runtime_error = loop {
+                break match watcher.into_future().compat().await {
```
[the remark this answers is not in the document]
Comment (ktff) [reply src/transforms/kubernetes/mod.rs]: We can't use while because we don't exit on that Error, but pass it to `WatchClient` [here](https://github.com/timberio/vector/blob/c044571879671bb0416a8cca12a3d3b829682baa/src/transforms/kubernetes/mod.rs#L162) that it then uses to construct next watch request.

File: src/transforms/kubernetes/mod.rs [hunk 9]
```diff
@@ -1 +1,552 @@
...
+                        self.metadata.refresh();
+                        continue;
+                    }
+                    Ok((None, _)) => RuntimeError::WatchUnexpectedlyEnded,
+                    Err((err, _)) => err,
+                };
+            };
+
+            warn!(
+                message = "Temporary stoped watching Pod metadata.",
```
[the remark this answers is not in the document]
Comment (ktff) [reply src/transforms/kubernetes/mod.rs]: Original message better describes what is happening because we aren't fetching it but watching it. On all, but one error, we won't get any metadata that we already have, once we resume watching.

File: src/transforms/kubernetes/mod.rs (hunk 1 above)
Comment (ktff) [reply src/transforms/kubernetes/mod.rs]: Because all `Fn` closures in there are different so we need to turn them into trait objects.

File: src/transforms/kubernetes/mod.rs (hunk 2 above)
Comment (ktff) [reply src/transforms/kubernetes/mod.rs]: Yes, we should.

File: src/transforms/kubernetes/mod.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (LucioFranco) [reply src/transforms/kubernetes/mod.rs]: the fluentd plugin has a cache_ttl option so it may just make sense to provide the ability to configure it.

File: src/transforms/kubernetes/mod.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (LucioFranco) [reply src/transforms/kubernetes/mod.rs]: I personally think if we are going to force uses to set a env var we should also allow it to be configurable via the toml config too.

File: src/transforms/kubernetes/mod.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (LucioFranco) [reply src/transforms/kubernetes/mod.rs]: Fair, but I also find it a bit odd that this fn returns an error without using a Result. I think to me that makes this much more clear to understand.

File: src/transforms/kubernetes/mod.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (LucioFranco) [reply src/transforms/kubernetes/mod.rs]: Should be possible to use `StreamExt::next` then and match internally.

[the remark this answers is not in the document]
Comment (LucioFranco) [reply src/transforms/kubernetes/mod.rs]: I think we should try to take advantage of stream here and while loops.

File: src/transforms/kubernetes/mod.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (LucioFranco) [reply src/transforms/kubernetes/mod.rs]: I think you can assume that the metadata watch stopped because it failed, right?

File: src/transforms/kubernetes/mod.rs (hunk 1 above)
Comment (LucioFranco) [reply src/transforms/kubernetes/mod.rs]: Fair, though I do find the usage of closures quite hard to follow. I wonder if there might be a better solution to extract a value from a struct like so.

**Comment by LucioFranco** [comment]
So I have two main questions:

1) How does using the watch client affect us, this means that we will need to have a tcp connection open to the api server for the duration of the vector instance. I don't think this should be a large issue but I have seen other metadata transforms provide the option to use the watch or a poll based api.

2) I think the biggest problem with large scale deployments would be vector trying to hit the api server all at once. We should introduce some sort of jitter for the initial request and subsequent requests so we can avoid the stampeding herd problem.

I think this follow closely to what you said @ktff, what do you think?

File: src/transforms/kubernetes/mod.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (ktff) [reply src/transforms/kubernetes/mod.rs]: True, I'll change it.

File: src/transforms/kubernetes/mod.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (ktff) [reply src/transforms/kubernetes/mod.rs]: Not necessarily, there are other reasons for reseting the connection besides a failure. For ex. server can say that we need to reset connection, or reset version. And I expect most of the times we will reset the connection for those reasons as they are normal to happen.

File: src/transforms/kubernetes/mod.rs (hunk 1 above)
Comment (ktff) [reply src/transforms/kubernetes/mod.rs]: One big `enum` with a variant for every supported field is also a solution. It should be clearer.

**Comment by ktff** [comment]
@LucioFranco 

Yep, we are thinking the same thing. I'll add the jitter. And we could open an issue for adding it to the retry policy, and possibly extract this implementation.

I suspect poll is useful for usecases when the used portion of metadata won't be changing, so only fetching it once makes sense. Other than that, I think we would need to wait and see. I also didn't found any Fluent issue regarding watch vs poll.

File: src/transforms/kubernetes/mod.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (ktff) [reply src/transforms/kubernetes/mod.rs]: True

**Comment by LucioFranco** [comment]
@binarylogic do you have anymore info you cold provide us about the issues you were describing?

File: src/transforms/kubernetes/mod.rs [hunk 10]
```diff
@@ -9,39 +9,69 @@ use crate::{
...
     #[serde(default = "default_fields")]
-    fields: Vec<String>,
+    fields: Vec<Field>,
+    /// For how long will we hold on to pod's metadata after the pod has been deleted.
+    /// seconds
+    #[serde(default = "default_cache_ttl")]
+    cache_ttl: u64,
+    /// Node name whose pod's metadata should be watched. Has priority
+    /// over value found in enviorment variable.
+    #[serde(default)]
```
Comment (LucioFranco) [inline src/transforms/kubernetes/mod.rs]: don't think you need the default attribute here.

File: src/transforms/kubernetes/mod.rs [hunk 11]
```diff
@@ -9,39 +9,69 @@ use crate::{
...
 #[derive(Deserialize, Serialize, Debug)]
 #[serde(deny_unknown_fields)]
 pub struct KubePodMetadata {
     #[serde(default = "default_fields")]
-    fields: Vec<String>,
+    fields: Vec<Field>,
+    /// For how long will we hold on to pod's metadata after the pod has been deleted.
+    /// seconds
+    #[serde(default = "default_cache_ttl")]
+    cache_ttl: u64,
```
Comment (LucioFranco) [inline src/transforms/kubernetes/mod.rs]: nit we've been using `Option` to mark a field as optional then setting the default when we need to unwrap it.

File: src/transforms/kubernetes/mod.rs [hunk 12]
```diff
@@ -67,12 +102,17 @@ impl TransformConfig for KubePodMetadata {
 
         // Run background task
         cx.executor().spawn_std(async move {
-            let error = metadata_client.run().await;
-            error!(message = "Stopped updating Pod metadata.", reason = %error);
+            match metadata_client.run().await {
+                Ok(()) => unreachable!(),
```
Comment (LucioFranco) [inline src/transforms/kubernetes/mod.rs]: I would think we probably just want to let this pass through, since technically a task could finish gracefully and just exit? (even if it doesn't work like this now thats how I would assume this type of api would work.

File: src/transforms/kubernetes/mod.rs [hunk 13]
```diff
@@ -103,108 +143,103 @@ enum BuildError {
...
+                        break;
+                    }
+                }
+            }
 
             warn!(
-                message = "Temporary stoped watching Pod metadata.",
-                reason = %runtime_error
+                message = "Temporarily stoped watching Pod metadata.",
+                reason = ?error
```
Comment (LucioFranco) [inline src/transforms/kubernetes/mod.rs]:
```suggestion
                reason = %error
```

This should be display if it impls Error

File: src/transforms/kubernetes/mod.rs [hunk 14]
```diff
@@ -288,124 +328,89 @@ impl Transform for KubernetesPodMetadata {
...
-    let key: Atom = with_prefix(name).into();
-    let fun = move |pod: &Pod| {
-        fun(pod)
-            .map(|data| vec![(key.clone(), data.into())])
-            .unwrap_or_default()
-    };
-    (name, Box::new(fun) as Box<_>)
-}
+    /// Extracts this field from Pod
+    fn extract(self, pod: &Pod) -> Option<Vec<(Atom, Value)>> {
```
Comment (LucioFranco) [inline src/transforms/kubernetes/mod.rs]: This is very easy to follow! :+1:

File: src/transforms/kubernetes/mod.rs [hunk 15]
```diff
@@ -288,124 +328,89 @@ impl Transform for KubernetesPodMetadata {
...
+        let prefix_key = Self::with_prefix(self.name()) + ".";
+
+        map.iter()
+            .map(|(key, value)| ((prefix_key.clone() + key).into(), value.into()))
+            .collect()
+    }
 
-fn with_prefix(name: &str) -> String {
-    event::log_schema().kubernetes_key().as_ref().to_owned() + "." + name
+    fn with_prefix(name: String) -> String {
```
Comment (LucioFranco) [inline src/transforms/kubernetes/mod.rs]: technically this could be a `&str` since we deref into `&name` below

File: src/transforms/kubernetes/mod.rs [hunk 16]
```diff
@@ -103,108 +143,103 @@ enum BuildError {
...
-                break match watcher.into_future().compat().await {
-                    Ok((Some(event), tail)) => {
-                        watcher = tail;
+            error = Some(RuntimeError::WatchUnexpectedlyEnded);
+            retry_timeout = self.max_retry_timeout.min(retry_timeout * 2);
 
+            while let Some(next) = watcher.next().await {
+                match next {
+                    Ok(event) => {
                         self.delete_update();
```
Comment (LucioFranco) [inline src/transforms/kubernetes/mod.rs]: What happens if we get no new updates, the only way for the cache to get invalidated is via a new event which then triggers cache garbage collection, right? Would it make sense to select on a timer here?

**Review by LucioFranco — CHANGES_REQUESTED** [review]
I'd like to see the config changes to Option and I'd also like to see us reenable kube tests before we merge. Otherwise, this looks fantastic

**Comment by binarylogic** [comment]
> @binarylogic do you have anymore info you cold provide us about the issues you were describing?

Not at this time. As long as we are cognizant of network congestion against this API that should be sufficient for a first version. A simple backoff algorithm should be sufficient for now.

File: src/transforms/kubernetes/mod.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (LucioFranco) [reply src/transforms/kubernetes/mod.rs]: Ok then the message should say "Temporarily unable to watch pod metadata."

**Comment by LucioFranco** [comment]
@binarylogic :+1: the new changes now provide a jitter to creating the initial connection and any subsequent ones.

**Comment by ktff** [comment]
@LucioFranco they are failing because of log file size which #1949 fixes. I'll merge it now.

File: src/transforms/kubernetes/mod.rs (hunk 12 above)
Comment (ktff) [reply src/transforms/kubernetes/mod.rs]: True.

File: src/transforms/kubernetes/mod.rs (hunk 15 above)
Comment (ktff) [reply src/transforms/kubernetes/mod.rs]: True

File: src/transforms/kubernetes/mod.rs (hunk 16 above)
Comment (ktff) [reply src/transforms/kubernetes/mod.rs]: Yes, that is the only way. And since that's also the only way for the cache to get bigger, using a timer won't make any more difference than freeing some kilobytes sooner rather than later.

Comment (LucioFranco) [reply src/transforms/kubernetes/mod.rs]: Well a timer would allow us to avoid issues with failing to connect to kube right? I am not sure I see a disadvantage to using a timer we do this in a few other places like batching.

Comment (ktff) [reply src/transforms/kubernetes/mod.rs]: What issues do you mean? Regarding postponing/never deleting entries of deleted Pod's, I don't see any way of them becoming a problem regardless of any issue.

File: src/transforms/kubernetes/mod.rs [hunk 17]
```diff
@@ -1 +1,568 @@
...
+            if found.is_none() {
+                warn!(
+                    message = "Metadata for pod not yet available.",
+                    pod_uid = ?std::str::from_utf8(pod_uid.as_ref()),
+                    rate_limit_secs = 10
+                );
+            }
+        } else {
+            warn!(
+                message = "Event without field, so it can't be enriched with metadata.",
```
Comment (LucioFranco) [inline src/transforms/kubernetes/mod.rs]: We should align these logs with what we have in other transforms:

https://github.com/timberio/vector/blob/master/src/transforms/json_parser.rs#L81

File: src/transforms/kubernetes/mod.rs [hunk 18]
```diff
@@ -1 +1,568 @@
...
+            )
+        } else {
+            "".to_owned()
+        };
+
+        CONFIG_MAP_YAML_WITH_METADATA.replace(FIELD_MARKER, replace.as_str())
+    }
+
+    #[test]
+    fn kube_metadata() {
```
Comment (LucioFranco) [inline src/transforms/kubernetes/mod.rs]: This seems like quite a small test, would it make sense to add some more coverage?

**Review by LucioFranco — COMMENTED** [review]
Things are looking good, just a few questions and I think we need to add docs!

File: src/transforms/kubernetes/mod.rs [hunk 19]
```diff
@@ -1 +1,568 @@
...
+    MissingNodeName { env: &'static str },
+    #[snafu(display("max_retry_timeout must be larger than 0."))]
+    TooSmallMaxRetryTimeout,
+}
+
+/// Vlient which watches for Pod metadata changes, extracts fields,
+/// and writes them to the metadata map.
+struct MetadataClient {
+    fields: Vec<Field>,
+    metadata: WriteHandle<Bytes, Box<(Atom, FieldValue)>>,
```
[the remark this answers is not in the document]
Comment (ktff) [reply src/transforms/kubernetes/mod.rs]: It is, `WriteHandle` requires that its value implement `ShallowCopy` which is unsafe to implement so we are using `Box` as a wrapper since `ShallowCopy` is implemented for it.

File: src/transforms/kubernetes/mod.rs (hunk 16 above)
Comment (ktff) [reply src/transforms/kubernetes/mod.rs]: I'll add it, then we can decide if the added complexity is worth it.

File: src/transforms/kubernetes/mod.rs (hunk 18 above)
Comment (ktff) [reply src/transforms/kubernetes/mod.rs]: Well, this test covers all the main paths except those of failure and deletion. Now deletion could theoretically be tested, but all the ways that I see are quite complex and definitely not worth it. 

There is also a separate test for `WatchClient`.

I think those cover it nicely.

File: src/transforms/kubernetes/mod.rs (hunk 17 above)
Comment (ktff) [reply src/transforms/kubernetes/mod.rs]: Something like this.

File: src/transforms/kubernetes/mod.rs [hunk 20]
```diff
@@ -1 +1,585 @@
...
+    #[snafu(display(
+        "Missing environment variable {:?} containing node name `spec.nodeName`.",
+        env
+    ))]
+    MissingNodeName { env: &'static str },
+    #[snafu(display("max_retry_timeout must be larger than 0."))]
+    TooSmallMaxRetryTimeout,
+}
+
+/// Vlient which watches for Pod metadata changes, extracts fields,
```
Comment (Hoverbear) [inline src/transforms/kubernetes/mod.rs]:
```suggestion
/// Client which watches for Pod metadata changes, extracts fields,
```

File: src/transforms/kubernetes/mod.rs [hunk 21]
```diff
@@ -1 +1,585 @@
...
+use k8s_openapi::api::core::v1::Pod;
+use rand::Rng;
+use serde::{Deserialize, Serialize};
+use snafu::{ResultExt, Snafu};
+use std::collections::{BTreeMap, VecDeque};
+use std::time::{Duration, Instant};
+use string_cache::DefaultAtom as Atom;
+use tokio::timer::Delay;
+
+// *********************** Defined by Vector **************************** //
```
Comment (Hoverbear) [inline src/transforms/kubernetes/mod.rs]: What's this all about?

File: src/transforms/kubernetes/mod.rs [hunk 22]
```diff
@@ -1 +1,585 @@
...
+
+/// Node name `spec.nodeName` of Vector pod passed down with Downward API.
+const NODE_NAME_ENV: &str = "VECTOR_NODE_NAME";
+
+/// 1h sounds a lot, but a metadata entry averages around 300B and in an extreme
+/// scenario when a pod is deleted every second, we would have ~1MB of probably
+/// not usefull data. Which is acceptable. The benefit of such a long delay
+/// is that we will certainly* have the metadata while we are processing the
```
Comment (Hoverbear) [inline src/transforms/kubernetes/mod.rs]: What's the `*` on the `certainly*`?

File: src/transforms/kubernetes/mod.rs [hunk 23]
```diff
@@ -1 +1,585 @@
...
+
+inventory::submit! {
+    TransformDescription::new_without_default::<KubePodMetadata>("kubernetes_pod_metadata")
+}
+
+#[typetag::serde(name = "kubernetes_pod_metadata")]
+impl TransformConfig for KubePodMetadata {
+    fn build(&self, cx: TransformContext) -> crate::Result<Box<dyn Transform>> {
+        // Main idea is to have a background task which will premptively
+        // acquire metadata for all pods on this node, and with it maintaine
```
Comment (Hoverbear) [inline src/transforms/kubernetes/mod.rs]:
```suggestion
        // acquire metadata for all pods on this node, and with it maintain
```

File: src/transforms/kubernetes/mod.rs [hunk 24]
```diff
@@ -1 +1,585 @@
...
+            Err(BuildError::TooSmallMaxRetryTimeout)
+        }
+    }
+
+    /// Listens for pod metadata changes and updates metadata map.
+    async fn run(mut self) -> Result<(), BuildError> {
+        let mut version = None;
+        let mut error = None;
+        // Since this transform will in most cases be deployed on all Nodes
+        // and fairly simulationisly, we can immediately start with user
```
Comment (Hoverbear) [inline src/transforms/kubernetes/mod.rs]:
```suggestion
        // and fairly simultaneously, we can immediately start with user
```

File: src/transforms/kubernetes/mod.rs (hunk 19 above)
[the remark this answers is not in the document]
Comment (LucioFranco) [reply src/transforms/kubernetes/mod.rs]: Ah ok I forgot there is a bytes feature and it does this for bytes :D

**Review by LucioFranco — APPROVED** [review]
I think just docs and a +1 @Hoverbear and we are good to merge!

File: src/transforms/kubernetes/mod.rs (hunk 21 above)
Comment (ktff) [reply src/transforms/kubernetes/mod.rs]: A leftover from times when configuration defined by Kubernetes was here. Now it's [here](https://github.com/timberio/vector/blob/6111a9c96a4460c0391e75be1d7b8d8f21b9d4ab/src/transforms/kubernetes/watch_client.rs#L18).

File: src/transforms/kubernetes/mod.rs (hunk 22 above)
Comment (ktff) [reply src/transforms/kubernetes/mod.rs]: `certainly` in all normal situations. An abnormal would be:
 - Node is run in a VM which was paused for an hour.
 - Large disk i/o and large amount of logs that need to be picked up, and ram overflowed to swap.
 - etc.

File: src/transforms/kubernetes/mod.rs [hunk 25]
```diff
@@ -1 +1,585 @@
...
+    #[snafu(display("max_retry_timeout must be larger than 0."))]
+    TooSmallMaxRetryTimeout,
+}
+
+/// Vlient which watches for Pod metadata changes, extracts fields,
+/// and writes them to the metadata map.
+struct MetadataClient {
+    fields: Vec<Field>,
+    metadata: WriteHandle<Bytes, Box<(Atom, FieldValue)>>,
+    /// (key of data to be deleted, can be deleted after this point in time)
```
[the remark this answers is not in the document]
Comment (ktff) [reply src/transforms/kubernetes/mod.rs]: Not sure if I'm following you, after the `Instant` it will be deleted while still in the queue and then removed from the queue.

**Comment by ktff** [comment]
There is one more issue that popped out during testing for the docs, keys of Kubernetes labels and annotations are usually constructed with a `.` which is causing for the key to be interpreted as nesting. 

For example:

label `"app.kubernetes.io/name" = value`
will be deconstructed as `{ "app" : { "kubernetes" : { "io/name" : value } } }` 

which isn't correct. And while it does behave as users would expect some times, some times it would behave differently, for example with any sink that outputs this as a `json`.

There are solutions for this, but I'm unsure which one is preferred:
 

1. Replace `.` with `_` in these keys. 
   - Pro, easy to implement.
   - Con, bad for UX. 
2. Add new `Value::Map` variant which isn't nestable.
   - Pro, would work as expected.
   - Con, unsure how much work is needed.  
3. Some other way of storing the key, without `.` being interpreted as nesting, that I'm unaware of.  

@LucioFranco @binarylogic @Hoverbear what do you think?

**Comment by ktff** [comment]
@binarylogic Initial documentation has been added.

**Comment by LucioFranco** [comment]
hmm, that is a big issue, I don't want to lose the value of having the keys contain `.` in them. @a-rodin since you did a lot of the new work around events do you have any thoughts around that? I assume since we nest now without leaving everything as flattened it should be possible to escape any of that nested dsl?

**Comment by binarylogic** [comment]
I'll be reviewing the docs this week.

**Comment by ktff** [comment]
@a-rodin could we get your thoughts on how to escape `.` being interpreted as nesting mentioned in before [comment](https://github.com/timberio/vector/pull/1888#issuecomment-595719639).

**Comment by ktff** [comment]
Once  #2082 is merged the '.' problem can be resolved and some documentation needs to be changed. 

Although #2081 is a solution for '.' problem, we should keep an eye on how error prone it will be for the users to remember putting '\\' on every '.' in a label that they want to use in `vector.toml`.

**Comment by Alexx-G** [comment]
This PR adds everything we need tn order to go with Vector in EKS! Thanks a ton!
I'm going to test today Vector 0.8.x for a small cluster, however for production use we really need the ability to filter by pod_name/namespace etc.
Is there a tiny chance this feature will be released with 0.9? No pressure, I just want to set correct expectations on my end.

**Comment by ktff** [comment]
@Alexx-G Great!

If you are using Kubernetes v1.14 or higher, the `kubernetes` source will output `pod_name` and `pod_namespace`. Although those names will be changed with the coming 0.9 version which will most likely have this transform in it.

**Comment by Alexx-G** [comment]
Yesterday I tested 0.8.x on an EKS cluster, and technically there's enough metainfo (contaienr_name, pod_name and namespace) so we can settle down with Vector. I tested it in a relatively small cluster, once it passes our validation phase, we'll see how it behaves for bigger clusters (100-200 worker nodes).
Since this feature makes API calls to kubernetes, the kubernetes deployment must have corresponding clusterrole bindings, right?

**Comment by binarylogic** [comment]
Thanks for all of your help @Alexx-G 

> Since this feature makes API calls to kubernetes, the kubernetes deployment must have corresponding clusterrole bindings, right?

@ktff if that is the case, woudl you mind updating https://github.com/timberio/vector/issues/1450 to include details on that. I'm working on the k8s docs now.

**Comment by Alexx-G** [comment]
@binarylogic Thank you for your effort! I'd gladly offer more help.
Today I'll whitelist more services that Vector will collect logs from and I'll see how it goes. But from my previous tests I don't expect any issues there.
Once I finish integrating Vector, I'll see if I can contribute with some feedback to #1397 (since it should contain all resources vector needs to work properly in a Kubernetes cluster).

**Comment by ktff** [comment]
@Alexx-G 

> Since this feature makes API calls to kubernetes, the kubernetes deployment must have corresponding clusterrole bindings, right?`

This transform yes, and its clusterrole binding is in the updated [config](https://github.com/timberio/vector/blob/28218fe58c8a89d4f985a009aceb8c9d645bbbd4/config/kubernetes/vector-daemonset.yaml). 


@binarylogic 

> if that is the case, woudl you mind updating #1450 to include details on that. I'm working on the k8s docs now.

`clusterrole bindings` are a part of `vector-daemonset.yaml`, and besides having RBAC authorization enabled, they don't require other user intervention. It is documented in the `yaml`, with other such necessities, so just mentioning `yaml` as the goto for more in depth information should be enough. 

 RBAC authorization is documented in this transforms documentation as it's the one using it. `source` can do without it.

**Comment by binarylogic** [comment]
Thank you. That makes sense.

**Comment by ktff** [comment]
@binarylogic could you review the docs.

**Comment by binarylogic** [comment]
Yep. I'm on it today. Once I finish the docs this is good to merge, correct?

**Review by binarylogic — APPROVED** [review]
Docs are done! Looks really good.

**Comment by binarylogic** [comment]
I am making a few more documentation changes and will merge afterwards.

**Comment by LucioFranco** [comment]
@ktff can we add a test around verifying labels are inserted correctly and all other values are inserted as we expect?

**Comment by binarylogic** [comment]
Nice work @ktff and @LucioFranco.

**Comment by marcbachmann** [comment]
Did this get removed from master because of performance issues?

**Comment by jszwedko** [comment]
Hi @marcbachmann . This was removed since it is implemented by the `kubernetes_logs` source. We have discussed bringing it back in the future though, to separate the reading of logs and annotation of metadata.
